docs/getting-started/tutorial.md

```diff
@@ -0,0 +1,349 @@
+
+The final question in our form is for the ZIP code, which is also the most complicated question. While ZIP codes are usually `string` data, they also need to follow specific formats. How do we specify data contraints in our `formConfig`?
+
+The primary way to do so is through the `schema` object. The JSON Schem specification includes multiple properties to add constraints to the data that is accepted for that field: `minLength`, `maxLength`, `format`, and `pattern`, among others.
+
+We're going to be using the `pattern` property to validate our ZIP code input, which takes a regex to match the user entry against.
+
+This is what our `schema` looks like with the ZIP code field added:
+```js
+...
+  schema: {
+    type: 'object',
+    properties: {
+      street: {
+        type: 'string'
+      },
+      city: {
+        type: 'string'
+      },
+      state: {
+        type: 'string',
+        enum: ['AK', 'IL', 'MA', 'TX', 'WY']
+      },
+      zip: {
+        type: 'string',
+        pattern: '^[0-9]{5,9}$'
+      }
+    }
+  },
+...
+```
+
+Wonderful! Now we have a ZIP code field. Let's test entering an invalid ZIP code. Enter an invalid entry and focus away from the ZIP code field:
+![ZIP code pattern error](../images/zip-pattern-error.png)
+
+Uh oh! While the form caught that the entry wasn't correct, and displayed an error to the user, the error isn't particularly readable.
+
+We can add custom error messages to our form for specific cases. Again, since this is related to what is displayed to the user, custom error messages go in the `uiSchema`:
+
+```js
+...
+  uiSchema: {
+    street: {
+      'ui:title': 'Street'
+    },
+    city: {
+      'ui:title': 'City'
+    },
+    state: {
+      'ui:title': 'State',
+      'ui:options': {
+        labels: {
+          AK: 'Alaska',
+          IL: 'Illinois',
+          MA: 'Massachusetts',
+          TX: 'Texas',
+          WY: 'Wyoming'
+        }
+      }
+    },
+    zip: {
+      'ui:title': 'ZIP code',
+      'ui:errorMessages': {
+        'pattern': 'Please enter a valid US ZIP code. It must be between 5 and 9 numbers.'
+      }
+    }
+  }
+...
+```
+
+In order to display a custom error message, we pass the `ui:errorMessages` object to `uiSchema` under the field for `zip`. The `ui:errorMessages` object takes key-value pairs, with the key being the name of the JSON Schema property the entry is in violation of, the value being the message we want displayed in that case. By doing this, we can add much more description error messages to the form to improve the user experience.
+
+Now we're seeing a much more useful error message:
+![Helpful error message](../images/helpful-error-message.png)
+
+
+## End of Tutorial
+
+Congratulations! You've now built your first form using only a JSON config file. Hopefully you are starting to see the potential benefits of being able to describe the form you want as opposed to having to build all of the components and UI patterns directly.
```

Comment:
> Congratulations! You've now built your first form using only a JSON config file. Hopefully you are starting to see the potential benefits of being able to describe the form you want as opposed to having to build all of the components and UI patterns directly.

> There are so many more things you can do with US Forms Systems, including much more complex form patterns, like conditionally expanded fields, custom validation functions, more complex form elements, groups of similar questions, and more. Continue reading through our [docs](../building-a-form/README.md) to learn more about what's possible!

What about:

"Well done! You've built your first form using only a JSON config file, and seen the benefits of describing your form rather than building components and UI patterns directly.

With the US Forms System, you can build more sophisticated form patterns, including conditionally expanded fields, custom validation functions, complex form elements, groups of similar questions, and more. Learn more about what's possible in "[Building a form](../building-a-form/README.md)!"